src/impl.cxx

```diff
@@ -1946,6 +1946,7 @@ namespace ipr {
               longlongtype(get_identifier("long long"), cxx_linkage(), anytype),
               ulonglongtype(get_identifier("unsigned long long"),
                             cxx_linkage(), anytype),
+              nullptrtype(get_identifier("nullptr_t"), cxx_linkage(), anytype),
```

Comment:
Why is `nullptr_t` is being made builtin type?  I am not necessarily saying it is wrong.  I am asking for rationale to understand the choice. From ISO C++ perspective, `nullptr_t` is actually defined in namespace `std` as an alias for `declttype(nullptr)`

Reply:
What should `declttype(nullptr)` return? In case it returns the very same typedef, it looks like infinite recursion.

Reply:
`decltype(nullptr)` does not need to return anything.

Reply:
Users might use `decltype(nullptr)` in their code, and we might need to generate warning messages that include the name of the type. Are you implying, the type represented by  `decltype(nullptr)`  would be called  `decltype(nullptr)` , without resolving it to anything else? I think this behavior might be surprising for some of the users.

Reply:
1.  Designating the type of `nullptr` by `decltype(nullptr)` is -- by definition -- exactly right.  There is nothing surprising about it.  That is its definition.

2.  If the user didn't include `<cstddef>` (and therefore couldn't utter `std::nullptr_t`), and wrote why would showing `nullptr_t` would not be confusing?

3. There is no type called `nullptr_t`, why would you want to show the user that type instead of the alias `std::nullptr_t` (assuming `<cstddef>` was included)?

Reply:
In C++, `decltype(foo)` is never a type, it resolves to a type. While I do understand your logic here, the users consuming the warning messages are not used to this reasoning. The reason might be familiarity as none of the toolchains I know about does what you suggest: https://godbolt.org/z/bnrfYa While I do understand that we do not want to copy existing toolchains, I suspect that the users grown to expect this behavior.

My other concern is the error-proneness of the approach. Someone working with IPR might (and I am pretty sure they will) assume that, it is possible to peel off levels of type aliases an reach a final underlying/canonical type. This is the mental model people working on C++ tools used to. Defining `decltype(nullptr)` with itself would break their mental invariant creating an endless loop in their code. This could alienate potential adopters.

While I do believe breaking the habit can be beneficial, what do we gain from the self-recursive definition? Is the gain worth to potentially alienate IPR-based tool authors who shoot themselves in their foot?

Forcing people to think outside of the box can be the right thing to do, but I'm a bit concerned about retention when we are doing this too often.

Reply:
> decltype(foo) is never a type

Wait, what?  Do you have a source reference for that?

> The reason might be familiarity as none of the toolchains I know about does what you suggest: https://godbolt.org/z/bnrfYa While I do understand that we do not want to copy existing toolchains, I suspect that the users grown to expect this behavior.

I don't get what exactly you are trying to illustrate with that example.

> My other concern is the error-proneness of the approach. Someone working with IPR might (and I am pretty sure they will) assume that, it is possible to peel off levels of type aliases an reach a final underlying/canonical type. This is the mental model people working on C++ tools are used to. Defining decltype(nullptr) with itself would break their mental invariant creating an endless loop in their code. This could alienate potential adopters.

You're confusing the semantic internal representation with the external display of message.

Reply:
> decltype(foo) is never a type

This was my interpretation of: https://eel.is/c++draft/dcl.type.decltype#1.6

> I don't get what exactly you are trying to illustrate with that example.

It is a reference what other implementations display in their messages. You can switch compilers to look at more.

> You're confusing the semantic internal representation with the external display of message.

No. At that point I was reasoning about the semantic representation. Let's imagine how we would implement a method that returns the underlying type of a typedef in both cases:
1. If `decltype(foo)` always resolves to a concrete type, we just repeat this type resolution in a loop until we have no `decltype` left.
2. Otherwise, if `decltype(foo) ` can resolve to `decltype(foo)` we need to check whether we reached fixed point, potentially compare whole subexpressions for equality that can have any form.

I'd argue the second one is harder to implement and harder to reason about and could alienate users of the IPR.

Reply:
> This was my interpretation of: https://eel.is/c++draft/dcl.type.decltype#1.6

That bullet does not say, does not imply, that `decltype(foo)` is not a type.

> It is a reference what other implementations display in their messages. You can switch compilers to look at more.

I saw what the compilers display. Nothing I saw in those messages invalidate what I said earlier.  If you believe it does, then you need to break your reasoning in smaller steps.

> At that point I was reasoning about the semantic representation.

Then the semantics of `decltype(nullptr)` is that it designates itself, e.g. `decltype(nullptr)`.  If you want it be something else, you need to provide logical steps to get there.

> Let's imagine how we would implement a method that returns the underlying type of a decltype in both cases:

You're making it more complicated than it is.  If you don't have `std::nullptr_t` declared in the current environment, there is no basis to show it to the user when they wrote `decltype(nullptr)` -- just display what they wrote and move on.

The more general point I would like to make is that `decltype` is a general operator, that by its very definition can be used to introduce new types and that is exactly what we used when `nullptr` was introduced in C++11.   Let's not jump the gun from unfamiliarity to not having adopters.  Let's keep the drama elsewhere.

Reply:
> The more general point I would like to make is that decltype is a general operator, that by its very definition can be used to introduce new types

I understand that point. The question is not about can, but about should. What is the advantage of one approach over the other?

> You're making it more complicated than it is

My argument is, if `decltype` can designate to itself, that complicates the implementation of code that deals with IPR. Not just warning messages, but other code in general. I demonstrated this with a hypothetical function that would return the canonical type for an expression, e.g.  for `decltype(int)` would return `int`. We either need to add a special case to such a function for every self-designating case or solve expression equality.  Since there is an implementation cost associated with such a representation, I was wondering what is the advantage of this representation that warrants paying this cost.

Reply:
> I understand that point. The question is not about can, but about should. What is the advantage of one approach over the other?

There is no builtin type `nullptr_t` in C++.  Why do you believe display a made-up builtin is better than `std::nullptr_t`?  Why do you need to display the raw internal IPR representation to the user?

> My argument is, if decltype can designate to itself, that complicates the implementation of code that deals with IPR.

That is a hypothesis that needs testing.

> I demonstrated this with a hypothetical function that would return the canonical type for an expression, e.g. for decltype(int) would return int. 

I don't understand.  What is `decltype(int)`?

> I was wondering what is the advantage of this representation that warrants paying this cost.

We should first represent the semantics, then worry about the presentation to the user later, then cost.  Worry about cost before getting the representation correct is premature optimization.

I don't know I can get you to accept that the type `decltype(nullptr)` is itself, as defined by the C++ standard semantics; I don't know I've seen argument to accept the made up type `nullptr_t`.

Reply:
> There is no builtin type nullptr_t in C++.

Implementation I checked treats them as a builtin, and I am not aware of any issues from this practice.

> Why do you need to display the raw internal IPR representation to the user?

I don't need to display that to the user, and actually I do not insist on this part of my argument. You convinced me that displaying `decltype(nullptr)` can be equally good. 

> That is a hypothesis that needs testing.

The hypothesis is based on the fact that trees/DAGs can be processed more easily than graphs with cycles. Usually, when we can get away without introducing cycles, I think we should. Cycles are, of course, sometimes necessary, or even advantegous. I'm verry happy with a self-referential `decltype(nullptr)` when I can see a practical advantage. This is why I was trying to push the discussion to hypothetical implementations, because I want to understand in what scenarios can this decision help the implementors.

> I don't understand. What is decltype(int)?

Sorry, that was a typo/thinko, I mean `decltype(int_var)`.

> We should first represent the semantics,

I totally argee. What we disagree on, is the semantics. In my view, it should always be possible, to get a canonical type of `decltype(foo)`, or similarly, it should be always possible to decide whether a type is canonical. I think having self-referential `decltype`s make the semantics of other operations, like deciding if a type is canonical more complex, and what do we gain? 

> I don't know I can get you to accept that the type decltype(nullptr) is itself, as defined by the C++ standard semantics; I don't know I've seen argument to accept the made up type nullptr_t.

It is ultimately your call. Since most of my concerns are implementation driven, I will not be able to provide you any evidence unless we start implementing, so I'm fine with going for your proposal and reevaluate this later on.

Reply:
The IPR type algebra is fully a graph; we should get comfortable with that.  I don't know what canonical types are, so I won't press that; I suspect that is probably a separate conversation.

In general, we should also get comfortable and familiar with the notion that there are irreducible (IPR) expressions.  They don't introduce any more complication in C++ or its semantics processing.

While I understand that it can help getting familiar with C++ semantics by looking at what other compilers are doing, I want to recommend against trying hard to make IPR looking exactly like the C++ compilers we are familiar with.  I would like to push for a far more regular and more general representation.  When actual problems surface we will analyze them and deal with them.  But, please let's not just add code because compiler XYZ does it that way.